### PR TITLE
Document Distributions DSL

### LINKED ISSUE #154: Document distributions DSL
The "distribution DSL feature" (https://github.com/probcomp/Gen/pull/103) was merged, but isn't yet added to the Gen documentation.

The PR contains a pretty detailed Markdown writeup about the feature.

We might only want to document those parts of the feature that are more stable / tested / complete, or we could explain the missing functionality that might be expected (gradients for vector-valued random choices is one).

### PR DESCRIPTION
**Pull request by bzinberg**
Documents the Distributions DSL by pasting in the PR description of https://github.com/probcomp/Gen/pull/103 and making light edits.

Fixes https://github.com/probcomp/Gen/issues/154.

## Tested
Previewed the doc site locally.

### CODE REVIEW

**Comment by ztangent** [comment]
@bzinberg thanks for doing this! i was almost going to do it myself. while you're at it, do you think you could add `geometric` to the current list of built-in distributions? for whatever reason, it's currently missing.

also, it seems to me like the note for the future shouldn't be included in the docs, maybe? that's more like an internal design note.

**Comment by bzinberg** [comment]
> @bzinberg thanks for doing this! i was almost going to do it myself. while you're at it, do you think you could add geometric to the current list of built-in distributions? for whatever reason, it's currently missing.

Let's call that a separate PR, so that submission is not delayed by (and so that I'm not on the hook for) anything beyond this one fix.

> also, it seems to me like the note for the future shouldn't be included in the docs, maybe? that's more like an internal design note.

Do we have a separate place for internal design notes (that people might ever actually read)?  I actually think it's quite reasonable to put a note like this in the docs (there are others that I deleted because they weren't as appropriate, as shown in the git diff) -- interested users may want to see and it's not secret.  WDYT?

**Comment by ztangent** [comment]
@bzinberg gotcha, just made a separate pull request just to make sure `geometric` is added to the docs. I think @marcoct and @alex-lew will have more informed thoughts on your second question!

File: docs/src/ref/distributions.md [hunk 1]
```diff
@@ -19,3 +17,207 @@ exponential
 laplace
 ```
 
+## Creating New Distributions via the Distributions DSL
+
+In addition to using the above built-in distributions, and to defining subtypes
+of `Gen.Distribution` in plain Julia, you may combine existing distributions to
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: You can't use the `@dist` DSL to combine existing distributions, unfortunately -- only one random choice per `@dist` :)

File: docs/src/ref/distributions.md [hunk 2]
```diff
@@ -19,3 +17,207 @@ exponential
...
+or
+```julia
+@dist function name(arg1, arg2, ..., argN)
+    body
+end
+```
+
+Here `body` is ordinary Julia code, with the constraint that `body` must
+contain exactly one random choice.  The value of the `@dist` expression is then
+a `Gen.distribution` object called `name`, parameterized by `arg1, ..., argN`,
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: `Gen.Distribution` with a capital `D`?

File: docs/src/ref/distributions.md [hunk 3]
```diff
@@ -19,3 +17,207 @@ exponential
...
+end
+```
+
+and think about the more natural random variable `:n_students`.  This leads to
+more natural inference programs, which can constrain and propose directly to
+the `:n_students` trace address.
+
+### Permitted means of combination
+
+Of course, it is not possible for `@dist` to work on any arbitrary `body`.  We
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: Maybe it is not obvious to readers of the docs -- so perhaps we can remove "Of course"?

File: docs/src/ref/distributions.md [hunk 4]
```diff
@@ -19,3 +17,207 @@ exponential
...
+   All expressions representing non-random values that depend on distribution
+   arguments are `ARG` expressions.
+3. `RND`: Random variables. All expressions whose runtime values may differ
+   across multiple calls to this distribution (with the same arguments) are
+   `RND` expressions.
+
+**Importantly, Julia control flow constructs generally expect `CONST` values:
+the condition of an `if` or the range of a `for` loop cannot be `ARG` or
+`RND`.**
+_(Developer note: We could change this: it would not be too difficult, e.g., to
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: I'd be in favor of removing the "developer note" here. If we want to make this a priority, we can open an issue?

File: docs/src/ref/distributions.md [hunk 5]
```diff
@@ -19,3 +17,207 @@ exponential
...
+a single direction of information flow -- anything that touches classified data
+is also classified. Also, Tabular v2's type system, which similarly separates
+deterministic and random values in its type system.)
+
+#### Restrictions
+Users may _not_ reassign to arguments (like `x` in the above example), and may
+not apply functions with side effects. Names bound to expressions of type `RND`
+must be used only once. e.g., `let x = normal(0, 1) in x + x` is not allowed.
+
+**TODO:** (Expand on scalar-vs.-vector problems here.)
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: Is it OK to leave this TODO in the docs?

File: docs/src/ref/distributions.md [hunk 6]
```diff
@@ -19,3 +17,207 @@ exponential
...
+2 + 3 :: CONST 						(by rule 4)
+x :: ARG 						(by rule 2)
+x * (2 + 3) :: ARG 					(by rule 5)
+exp(x) :: ARG 						(by rule 5)
+normal(exp(x), exp(x)) :: RND 				(by rule 3)
+log(normal(exp(x), exp(x))) :: RND 			(by rule 6)
+log(normal(exp(x), exp(x))) + (x * (2 + 3)) :: RND 	(by rule 6)
+```
+
+## A note for the long-term future
```
Comment (alex-lew) [inline docs/src/ref/distributions.md]: What do you see as the benefit for putting this in the Gen documentation?

**Comment by alex-lew** [comment]
Thanks for doing this, Ben! Left a few comments but overall looks good.

File: docs/src/ref/distributions.md (hunk 4 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: Good call, an issue is the right place for this.  It's also probably the right place for the "long-term future note" below.  Filed https://github.com/probcomp/Gen/issues/199.

File: docs/src/ref/distributions.md (hunk 5 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: Issue would be better -- filed https://github.com/probcomp/Gen/issues/200, thanks.

File: docs/src/ref/distributions.md (hunk 6 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: The benefit was that users (who may be researchers in this area) can be aware of the discussion and participate if they want.  But as you pointed out above, an issue is a better place for this.  Filed https://github.com/probcomp/Gen/issues/201.

File: docs/src/ref/distributions.md (hunk 3 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: Done

File: docs/src/ref/distributions.md (hunk 2 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: Done

File: docs/src/ref/distributions.md (hunk 1 above)
Comment (bzinberg) [reply docs/src/ref/distributions.md]: Fixed.

**Review by bzinberg — COMMENTED** [review]
Good calls on all fronts, I think I've addressed them all

**Comment by marcoct** [comment]
@bzinberg This looks great, but I have one comment:

Users may implement new distributions by implementing the Distribution interface directly, or with the Distributions DSL. Currently, implementing new distributions by implementing the Distribution interface directly is documented under 'Extending Gen' here: 

https://probcomp.github.io/Gen/dev/ref/extending/#Custom-distributions-1

The title of the section ' ## Creating New Distributions via the Distributions DSL' might lead some readers of the 'Probability Distributions' section to believe that the distributions DSL is the only way to create a new distribution. It would be good to either add a link to the 'Extending Gen' section from this 'Probability Dsitributions' section, or to move the Distributions DSL documentation to Extending Gen.

**Comment by marcoct** [comment]
@bzinberg Ah I see you mentioned it -- maybe just include a link in that first sentence to https://probcomp.github.io/Gen/dev/ref/extending/#Custom-distributions-1.

**Comment by alex-lew** [comment]
@marcoct @bzinberg Maybe it makes sense to cross-reference in both places? Or, as you suggested, put them booth in the same place. 

We could also be more explicit about the different use cases? "Gen can be extended with new probability distributions, which can then be used to make random choices within generative functions.  Simple transformations of existing distributions can be created using the [Distributions DSL]. Users can construct more complex custom distributions by implementing Gen's `Distribution` interface directly."

**Comment by bzinberg** [comment]
@marcoct @alex-lew Yeah, that totally makes sense.  I took a crack at incorporating your suggestions, WDYT?

**Comment by marcoct** [comment]
@bzinberg Looks great!

**Comment by bzinberg** [comment]
Thanks @alex-lew for writing [basically everything](https://github.com/probcomp/Gen/pull/197/commits/ba26d2bf57f79b1a9efd1bedf9abb8d524b0b4b1), so that all we had to do here were minor [touch-ups](https://github.com/probcomp/Gen/pull/197/files/ba26d2bf57f79b1a9efd1bedf9abb8d524b0b4b1..862c32ed9aeb5a86660c3b6c5e4a8168839b762a).

**Pull request merged**